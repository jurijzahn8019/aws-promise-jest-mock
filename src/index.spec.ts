/* eslint-disable @typescript-eslint/no-explicit-any */
import { SecretsManager } from "aws-sdk";
import { on, infer } from ".";

jest.mock("aws-sdk");

describe("aws-mock", () => {
  it("Should create resolve mock from type", async () => {
    const m = on(SecretsManager)
      .mock("getSecretValue", infer)
      .resolve({ SecretString: "foo-bar" });

    const res = new SecretsManager({
      credentials: { accessKeyId: "rootkeyfoo" } as any
    })
      .getSecretValue({ SecretId: "bar-baz" })
      .promise();

    await expect(res).resolves.toMatchSnapshot("Result");
    expect(m.mock).toHaveBeenCalledTimes(1);
    expect(m.and.serviceMock.mock.calls[0]).toMatchSnapshot("Constructor");
  });

  it("Should create reject mock from type", async () => {
    const m = on(SecretsManager)
      .mock("getSecretValue", infer)
      .reject("foo-baz");

    const res = new SecretsManager()
      .getSecretValue({ SecretId: "bar-baz" })
      .promise();

    await expect(res).rejects.toMatchSnapshot("Result");
    expect(m.mock).toHaveBeenCalledTimes(1);
  });

  it("Should create resolve mock from instance", async () => {
    const m = on(new SecretsManager())
      .mock("getSecretValue", infer)
      .resolve({ SecretString: "foo-bar" });

    const res = m.service.getSecretValue({ SecretId: "bar-baz" }).promise();

    await expect(res).resolves.toMatchSnapshot("Result");
    expect(m.mock).toHaveBeenCalledTimes(1);
  });

  it("Should create reject mock from instance", async () => {
    const m = on(new SecretsManager())
      .mock("getSecretValue", infer)
      .reject(Error("baz-BAR"));

    const res = m.service.getSecretValue({ SecretId: "bar-baz" }).promise();

    await expect(res).rejects.toMatchSnapshot("Result");
    expect(m.mock).toHaveBeenCalledTimes(1);
  });

  it("Should not create snapshot", async () => {
    const m = on(SecretsManager)
      .mock("getSecretValue", infer)
      .resolve({ SecretString: "foo-bar" }, { snapshot: false });

    const res = new SecretsManager()
      .getSecretValue({ SecretId: "bar-baz" })
      .promise();

    await expect(res).resolves.toMatchSnapshot("Result");
    expect(m.mock).toHaveBeenCalledTimes(1);
  });

  it("Should invoke result callabck", async () => {
    const m = on(SecretsManager)
      .mock("getSecretValue", infer)
      .resolve(
        () => {
          return {
            SecretString: "FOO"
          };
        },
        { snapshot: false }
      );

    const res = new SecretsManager()
      .getSecretValue({ SecretId: "bar-baz" })
      .promise();

    await expect(res).resolves.toMatchSnapshot("Result");
    expect(m.mock).toHaveBeenCalledTimes(1);
  });

  it("Should chain mocks", async () => {
    const m = on(SecretsManager, { snapshot: false })
      .mock("getSecretValue", infer)
      .resolve(() => {
        return {
          SecretString: "FOO"
        };
      })
      .and.mock("createSecret", infer)
      .resolve({ Name: "FOO_SECRET" });

    const smm = new SecretsManager();
    smm.getSecretValue({ SecretId: "bar-baz" }).promise();

    await expect(
      smm.getSecretValue({ SecretId: "bar-baz" }).promise()
    ).resolves.toMatchSnapshot("getSecretValue");

    await expect(
      smm.createSecret({ Name: "FOO", SecretString: "BAR" }).promise()
    ).resolves.toMatchSnapshot("createSecret");

    expect(m.serviceMockBuilder.serviceMock).toHaveBeenCalledTimes(1);
  });
});
