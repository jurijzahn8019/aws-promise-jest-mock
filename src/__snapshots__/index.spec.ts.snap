// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`aws-mock Should chain mocks: createSecret 1`] = `
Object {
  "Name": "FOO_SECRET",
}
`;

exports[`aws-mock Should chain mocks: getSecretValue 1`] = `
Object {
  "SecretString": "FOO",
}
`;

exports[`aws-mock Should create reject mock from instance: Result 1`] = `[Error: baz-BAR]`;

exports[`aws-mock Should create reject mock from instance: secretsmanager.getSecretValue (reject mock 3b67f997b83b065e32ad7fb2b0a969e8d1dd7a2b) 1`] = `
Array [
  Object {
    "SecretId": "bar-baz",
  },
]
`;

exports[`aws-mock Should create reject mock from type: Result 1`] = `[Error: foo-baz]`;

exports[`aws-mock Should create reject mock from type: secretsmanager.getSecretValue (reject mock 3b67f997b83b065e32ad7fb2b0a969e8d1dd7a2b) 1`] = `
Array [
  Object {
    "SecretId": "bar-baz",
  },
]
`;

exports[`aws-mock Should create resolve mock from instance: Result 1`] = `
Object {
  "Account": "foo",
  "Arn": "arn:foo",
  "UserId": "fooser",
}
`;

exports[`aws-mock Should create resolve mock from instance: sts.getCallerIdentity (resolve mock 97d170e1550eee4afc0af065b78cda302a97674c) 1`] = `Array []`;

exports[`aws-mock Should create resolve mock from type: Constructor 1`] = `
Array [
  Object {
    "credentials": Object {
      "accessKeyId": "rootkeyfoo",
    },
  },
]
`;

exports[`aws-mock Should create resolve mock from type: Result 1`] = `
Object {
  "SecretString": "foo-bar",
}
`;

exports[`aws-mock Should create resolve mock from type: secretsmanager.getSecretValue (resolve mock 3b67f997b83b065e32ad7fb2b0a969e8d1dd7a2b) 1`] = `
Array [
  Object {
    "SecretId": "bar-baz",
  },
]
`;

exports[`aws-mock Should invoke result callabck: Result 1`] = `
Object {
  "SecretString": "FOO",
}
`;

exports[`aws-mock Should not create snapshot: Result 1`] = `
Object {
  "SecretString": "foo-bar",
}
`;
